File: C++/代码随想录/Hashtable/1. Two Sum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        num_dict = {}
        for i, num in enumerate(nums):
            complement = target - num
            if complement in num_dict:
                return [num_dict[complement], i]
            num_dict[num] = i
        return []

sol = Solution()

# 测试用例
print(sol.twoSum([2, 7, 11, 15], 9))  # 应该输出 [0, 1]
print(sol.twoSum([3, 2, 4], 6))  # 应该输出 [1, 2]
print(sol.twoSum([3, 3], 6))  # 应该输出 [0, 1]

File: C++/代码随想录/Hashtable/1. Two Sum.cpp
// 1. Two Sum
// https://leetcode.com/problems/two-sum/

#include <iostream>
#include <vector>
#include <unordered_map>

class Solution {
public:
    std::vector<int> twoSum(std::vector<int>& nums, int target) {
        std::unordered_map<int, int> numToIndex;
        for (int i = 0; i < nums.size(); ++i) {
            int complement = target - nums[i];
            if (numToIndex.find(complement) != numToIndex.end()) {
                return {numToIndex[complement], i};
            }
            numToIndex[nums[i]] = i;
        }
        return {};  // 返回空数组，实际上这种情况在题目中不会发生，因为题目保证有解。
    }
};

int main() {
    Solution sol;
    std::vector<int> nums = {2, 7, 11, 15};
    int target = 18;
    std::vector<int> res = sol.twoSum(nums, target);
    for (int i : res) {
        std::cout << i << " ";
    }
    std::cout << std::endl;

    return 0;
}
